raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1514,19 +1527,22 @@ private CompletableFuture<FetchResponseData> handleFetchRequest(
             || FetchResponse.recordsSize(partitionResponse) > 0
             || request.maxWaitMs() == 0
             || isPartitionDiverged(partitionResponse)
-            || isPartitionSnapshotted(partitionResponse)) {
+            || isPartitionSnapshotted(partitionResponse)
+            || isHighWatermarkUpdated(partitionResponse, fetchPartition)) {
             // Reply immediately if any of the following is true
             // 1. The response contains an error
             // 2. There are records in the response
             // 3. The fetching replica doesn't want to wait for the partition to contain new data
             // 4. The fetching replica needs to truncate because the log diverged
             // 5. The fetching replica needs to fetch a snapshot
+            // 6. The fetching replica should update its high-watermark
```

Comment:
Not specific to this PR, but when debugging deferred requests in the past, I always wished for some trace logging to enable to see why things were being completed. For example, if request CPU is really high and things are not being deferred as expected, it could be due to a too low maxWaitMs or any of these other reasons. Maybe something to consider adding down the road.